File: 876-hand-of-straights/hand-of-straights.py
class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        
        d = {}
        pq = []
        for i in hand:
            if i not in d:
                d[i]= 1
                heapq.heappush(pq,i)
            else:
                d[i]+=1
        
        while pq:
            x = pq[0]
            for i in range(groupSize):
                y = x + i
                if y not in d: return False
                d[y]-=1
                if d[y]==0:
                    d.pop(y)
            while pq and pq[0] not in d:
                # d.pop(pq[0])
                heapq.heappop(pq)
        return True

File: 876-hand-of-straights/hand-of-straights.cpp
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        unordered_map<int,int> d;
        priority_queue<int,vector<int>,greater<int>> pq;

        for(int i : hand){
            d[i]++;
            if(d[i]==1){
                pq.push(i);
            }
        }

        while(pq.size()>0){
            int x = pq.top();
            for(int i = 0;i<groupSize;i++){
                int y = x +i;
                if(d.count(y)==0) return false;
                d[y]-=1;
                if(d[y]==0){
                    d.erase(y);
                }
            }

            while (pq.size()>0 && d.find(pq.top())==d.end()){
                pq.pop();
            }
        }
        return true;
    }
};